arch/x86/kvm/pkvm/mmu.c

```diff
@@ -612,3 +624,115 @@ int pkvm_hyp_donate_host_mmio_locked(unsigned long phys, unsigned long size)
 
 	return pkvm_pgtable_map(&host_mmu, phys, phys, size, prot);
 }
+
+/**
+ * pkvm_host_share_hyp() - Shares pages from host to hypervisor.
+ * @phys:	Physical address of the memory region to share.
+ * @size:	Size of the memory region to share.
+ *
+ * The sharing transfers the memory pages in range [@phys, @phys + @size) which
+ * are exclusively owned by the host to shared-owned by the host and the pKVM
+ * hypervisor. The host_share_hyp_count counter of those memory pages are
+ * incremented. If a memory page is already shared-owned, page state will not be
+ * changed but only increment the counter. @phys and @size are not required to be
+ * PAGE_SIZE aligned (@size is required to be non-zeroed value) to support
+ * sharing page-unaligned memory pages.
+ *
+ * Returns: 0 on success, or a negative error code on failure.
+ */
+int pkvm_host_share_hyp(unsigned long phys, unsigned long size)
+{
+	unsigned long end = PAGE_ALIGN(phys + size);
+	enum pkvm_page_state pgstates[] = {
+		PKVM_PAGE_OWNED,
+		PKVM_PAGE_SHARED_OWNED,
+	};
+	struct pkvm_page *page;
+	int ret;
+
+	if (size == 0)
+		return -EINVAL;
+
+	pkvm_host_mmu_lock();
+
+	ret = check_host_mem_pgstates(phys, size, pgstates, ARRAY_SIZE(pgstates));
+	if (ret)
+		goto unlock;
+
+	/*
+	 * TODO: Prevent sharing a memory page which is
+	 * shared-owned with a guest VM.
+	 */
+
+	for (phys = PAGE_ALIGN_DOWN(phys); phys < end; phys += PAGE_SIZE) {
+		page = pkvm_phys_to_page(phys);
+
+		if (page->host_state == PKVM_PAGE_OWNED)
+			page->host_state = PKVM_PAGE_SHARED_OWNED;
+
+		/*
+		 * The page state should be shared-owned otherwise
+		 * it must be a code bug.
```

Comment:
nit: this comment is obvious?

Reply:
This is addressed by https://github.com/intel-staging/pKVM-IA/pull/43#discussion_r2484928456